Builtin/profiler: Replace builtin profiler

t4im opened the pull request:
Mod Profiler rewrite/merge of [profiler](https://github.com/t4im/profiler/) to replace the current minetest profiler.
Tracking issue #3849 

#### Changes
* Correct calculation of average (mean) values. (Also made undistorted by idleness.)
* Reduce instrumentation overhead.
* Updated and extended instrumentation
* Fix crashes related to missing parameters for the future. *(e.g. #4220)*
* Fix occasional DIV/0's.
* Prevent issues caused by timetravel (overflow, timejump, NTP corrections)
* Prevent modname clashes with internal names.
* Add `/profiler` command for profile-data management.
* Measure each instrumentation individually and label based on registration order.
* Allow labeling of ABM's and LBM's for easier classification.  
  Giving several ABM's or LBM's the same label will treat them as one.  
  Missing labels will be generated based on name or registration order.  
  *I've already added some labels to homedecor, pipeworks and technic last year; so you can already see that in action.*
* Improved table formatting.
* Measure profiler overhead
* Very configurable with settings menu integration.  
  (E.g. Skip instrumenting builtin)


#### Usage
Use the setting "profiler.load" to load the profiler at startup.

* `/profiler dump [filter]` - report statistics to STDOUT and debug.txt
* `/profiler print [filter]` - report statistics to in-game console
* `/profiler save [format [filter]]` - saves the runtime statistics to a file in your worldpath (`profile-<time>[-<filter>].<format>`)
  * `txt` (default) - same treetable format as used by the dump and print commands
  * `csv` - ready for spreadsheet import
  * `json` - useful for adhoc D3 visualizations
  * `json_pretty` - line wrapped and intended json for humans
  * `lua` - serialized lua table of the profile-data, for adhoc scripts  
* `/profiler reset` - reset all gathered profile data.
  This can be helpful to discard of any startup measurements that often spike during loading or to get more useful min-values.

 `[filter]` allows limiting the output of the data via substring/pattern matching against the modname.
Note: Serialized data structures might be subject to change with changed or added measurements.  
	csv might be the most stable, due to flat structure.

I've got some incomplete changes in stash, that i kept back to not mess with the stability of this version here. But if people are interested, i might add these at a later date (things like node instrumentation, call frequency, problem markers).

Fixer-007 commented:
This is very needed for debug

est31 commented on doc/lua_api.txt:
```diff
@@ -3402,6 +3405,9 @@ Definition tables
 ### LBM (LoadingBlockModifier) definition (`register_lbm`)
 
     {
+		label = "Upgrade legacy doors",
```
Don't use tabs in lua_api.txt.

est31 commented on builtin/profiler/init.lua:
```diff
@@ -0,0 +1,42 @@
+	})
+end
+
+core.log("action", "Initialize mod profiler")
```
Is this needed?

est31 commented on builtin/profiler/sampling.lua:
```diff
@@ -0,0 +1,150 @@
+end
+
+---
+-- Log a measurment for the sampler to pick up later.
```
Is it really spelled "measurment"? I don't think it is

Zeno- commented on builtin/profiler/reporter.lua:
```diff
@@ -0,0 +1,196 @@
+	end
+	local content, err = serialize_profile(profile, format, filter)
+	if not content then
+		return false, "Saving of profile failed with: " .. err
```
Shouldn't `output` be closed before returning?

t4im replied:
Ah, yes, thanks. :-)

est31 commented on builtin/profiler/sampling.lua:
```diff
@@ -0,0 +1,149 @@
+--
+local function sample(dtime)
+	-- Rare, but happens and is currently of no informational value.
+	if logged_time == 0 then return end
```
@kwolekr doesn't like having the then part of an if clause on the same line.

est31 commented:
Other than that style issue (which appears to be existing in other parts as well), I'd give it a :+1:

t4im commented:
Style issue is addressed. I'll pay attention to that next time.

t4im commented:
When people copy profiler tables into github, it shouldn't mess up things so much anymore:

 instrumentation                                         |    min µs |    max µs |    avg µs | min % | max % | avg %
-------------------------------------------------------- | --------- | --------- | --------- | ----- | ----- | ------
 unified_inventory:                                      |         1 |     10631 |        41 |   0.3 |  75.8 |  24.2
  - globalstep[1] .....................................  |         1 |       191 |        15 |   0.0 |  75.8 |  24.0


Not very pretty, but still better than the alternative non-monospace amagedon.

Zeno- commented:
I like it!

est31 commented:
me too!

Zeno- commented:
+1

t4im commented:
Rebased against master, squashed a few commits, edited a few messages.
No real changes since the last approval, [but a tiny fix found during testing](https://github.com/minetest/minetest/pull/4245/commits/7db4cc4bc9b8dad700ee4fc621fa32ecd31db75f#diff-c352a1dd10c1b6133672bd3dd435d984R109) (extra 'q' to not lose header entry)

est31 commented on builtin/settingtypes.txt:
```diff
@@ -1163,6 +1154,51 @@ secure.trusted_mods (Trusted mods) string
+
+#    The default format in which profiles are being saved,
+#    when calling `/profiler save [format]` without format.
+profiler.default_report_format (Default report format) enum txt,csv,lua,json,json_pretty
```
Maybe add a default for this enum?

est31 commented:
+1
